### PR TITLE
Handle pointer click events with multiple sources

### PR DESCRIPTION
**Pull request by rytis-buz**
Handle pointer click events that come from multiple input sources at the same time

### Description

It's possible that OS will produce two identical events with different source types when making a tap/click. 
E.g. On Samsung Dex with a mouse attached the OS will produce identical Touch events and Mouse events at the same time.
Until now this situation was not handled and each event would cancel each other out as we would remove all Pointer Statuses of different type than the one we are currently processing. So it went like this:
1. Touch ACTION_DOWN received - create a pointer status of type TOUCH
2. Mouse ACTION_DOWN received - find all pointer statuses with different types and remove them (removes previous TOUCH), create a pointer status of type MOUSE
3. Touch ACTION_UP received - find all pointer statuses with different types and remove them (removes previous MOUSE)
4. Mouse ACTION_UP received - find all pointer statuses with different types and remove them (removes previous TOUCH), create a pointer status of type MOUSE
So there is never a UP/DOWN action register in any Pointer status as they keep getting removed

### Changes made

Now the pointers are not disregarded during event collection, this was moved to event Processing state. During processing we find the first Pointer that has changed in the frame and we only process inputs from that Pointer type, disregarding other types. This prevents the Pointer type flip-flopping and Pointers canceling each other out

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [x] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Comment by rytis-buz** [comment]
Investigating test failures...

**Comment by rytis-buz** [comment]
Removing reviewers for now, need to investigate a bit more

**Comment by lyndon-unity** [comment]
I'm investigating an issue which seems very similar and this PR fixes that case. I'll link the bugs
https://jira.unity3d.com/browse/ISXB-342

**Comment by rytis-buz** [comment]
@lyndon-unity Could you test if your issue is fixed by using this branch?
I'm still investigating what the behaviour should with `SingleUnifiedPointer` as I get weird results, but it seems it was the same before my PR

**Comment by rytis-buz** [comment]
Sorry, I missed that it fixes the issue, disregard my comment

**Comment by lyndon-unity** [comment]
> Sorry, I missed that it fixes the issue, disregard my comment
The latest version still fixes the issue I was seeing. :) 

There is some tidy up needed from the last commit though as ProcessPointerButton no longer needs the state passing through and you can now remove this I think : public PointerEventData.FramePressState framePressState => m_FramePressState;

PS I'm seeing this warning which would be nice to remove 
InputSystem\InputManager.cs
The variable 'dropStatusEvents' is assigned but its value is never used

By moving 
bool dropStatusEvents = false;
inside
#if UNITY_EDITOR

**Comment by rytis-buz** [comment]
Yes, cleanup incoming, thanks

**Comment by lyndon-unity** [comment]
Test failures are due to the input team needing to merge down this branch into develop
dmytro/common-input-api-for-ui

https://github.com/Unity-Technologies/InputSystem/compare/dmytro/common-input-api-for-ui

This enabled the Input System back end for the new 'Input for UI' layer (which the UI TK interfaces to) that has landed on trunk.
I'll discuss merging this down with the team tomorrow.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (lyndon-unity) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: Could also list this bug number,
https://jira.unity3d.com/browse/ISXB-342

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs
Comment (lyndon-unity) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs:2944]: Thanks - good move. Not sure about change from bool to var, but I don't have string feeling on that as its obvious from the assignment.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs
Comment (lyndon-unity) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs]: Could be nice to pull this out into separate function to call to shrink the size of the main function.

File: Assets/Tests/InputSystem/Plugins/UITests.cs
Comment (lyndon-unity) [inline Assets/Tests/InputSystem/Plugins/UITests.cs]: I'm not 100% sure what this means here - is this expected or are we left with a remaining bug?

**Review by lyndon-unity — APPROVED** [review]
Change looks good and fixed the case I was investigating.

Great to see test addition, one small query on the test case for SingleUnifiedPointer

Minor feedback suggestions on splitting out a function and adding reference to extra bug but happy to approve

**Review by lyndon-unity — APPROVED** [review]
Change looks good and fixed the case I was investigating.

Great to see test addition, one small query on the test case for SingleUnifiedPointer

Minor feedback suggestions on splitting out a function and adding reference to extra bug but happy to approve

Would like to see tests run again with develop merged when we merge branch listed above to fix some test runs.

**Comment by jfreire-unity** [comment]
> Test failures are due to the input team needing to merge down this branch into develop dmytro/common-input-api-for-ui
> 
> https://github.com/Unity-Technologies/InputSystem/compare/dmytro/common-input-api-for-ui
> 
> This enabled the Input System back end for the new 'Input for UI' layer (which the UI TK interfaces to) that has landed on trunk. I'll discuss merging this down with the team tomorrow.

The trunk tests are not required for this PR to be merged. But was good to know that we need to fix some issues with InputForUI module work.
I re-run the required failed tests (that failed due to timeouts 🤷‍♂️) and now all the required tests pass and the PR can be merged ✅

File: Assets/Tests/InputSystem/Plugins/UITests.cs
Comment (rytis-buz) [reply Assets/Tests/InputSystem/Plugins/UITests.cs]: Honestly I don't know. I was looking forward for some comments on this. I've tested `SingleUnifiedPointer` in real life scenario and it works fine.
This test case produced the same results before my changes as well, also this code produced the same result:
```   
        BeginTouch(0, firstPosition, screen: touch1, queueEventOnly: true);
        yield return null;
        EndTouch(0, firstPosition, screen: touch1, queueEventOnly: true);
        yield return null;
```
The above situation just mimics a simple click with touch, yet produces Drop as well.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs
Comment (rytis-buz) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModule.cs]: Done cc71b58cf68651ce112592b836f2026a2e57587e

**Review by lyndon-unity — APPROVED** [review]
Nice, thank you.

**Review by robertas-unity — APPROVED** [review]
Finished testing on iPad - looks good.

Both touch & mouse input events are correct, no additional clicks are registered, other touch info (radius, position, deltas, etc.) are also correct.

Checked various IMGUI, Unity UI & UI Toolkit elements (buttons, text fields, toggles, etc.) - all of them respond to both touches and mouse clicks.

Tested with:
iPad Mini 6, iPadOS 15.0
iPad Pro, iPadOS 16.4.1

**Review by IGuscin — APPROVED** [review]
Works fine, no PR related issues were found.
Tested on:
Pixel 7 Android 13
Samsung Galaxy Z Fold2 5G Android 11 (both standalone and using Samsung Dex)
Intel Comet Lake U Chromebook Android 11
HP Chromebook 14A G5 Android 9

**Comment by lyndon-unity** [comment]
2022.1 test failures can be ignored as this is no longer a supported Unity version and we will remove the CI checks soon.

**Comment by stefanunity** [comment]
Paulius & I finished testing. No new issues found. https://docs.google.com/document/d/14zx24BQgCKC8rVvzQlh-j4G1-gfTUymmKiFyQwwZxGw/edit#

Cases requiring specific hardware have been verified by mobile QA folks.

Good to merge.

**Comment by rytis-buz** [comment]
@jamesmcgill @jfreire-unity PR is missing your approvals
Also, who's authorised to merge it? :)

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (jamesmcgill) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: This will need moved back into [Unreleased] now after merges/release

**Review by jamesmcgill — APPROVED** [review]
Looks good, thanks!
Just need to fixup the CHANGELOG again.

**Comment by jamesmcgill** [comment]
Thanks @rytis-buz!
Merged now.

**Pull request merged**
